plugins/draco/io/DracoWriter.hpp

```diff
@@ -0,0 +1,159 @@
+        { pdal::Dimension::Id::Z,         draco::GeometryAttribute::POSITION },
+        { pdal::Dimension::Id::NormalX,   draco::GeometryAttribute::NORMAL },
+        { pdal::Dimension::Id::NormalY,   draco::GeometryAttribute::NORMAL },
+        { pdal::Dimension::Id::NormalZ,   draco::GeometryAttribute::NORMAL },
+        { pdal::Dimension::Id::Red,       draco::GeometryAttribute::COLOR },
+        { pdal::Dimension::Id::Green,     draco::GeometryAttribute::COLOR },
+        { pdal::Dimension::Id::Blue,      draco::GeometryAttribute::COLOR },
+        { pdal::Dimension::Id::TextureU,  draco::GeometryAttribute::TEX_COORD },
+        { pdal::Dimension::Id::TextureV,  draco::GeometryAttribute::TEX_COORD },
+        { pdal::Dimension::Id::TextureW,  draco::GeometryAttribute::TEX_COORD }
+    };
+
+};
+
+namespace pdal
+{
+    typedef std::shared_ptr<std::ostream> FileStreamPtr;
+
+class PDAL_DLL DracoWriter : public Writer/*, public Streamable*/
+{
+public:
+
+    DracoWriter();
+    ~DracoWriter();
+    std::string getName() const;
+private:
+    virtual void addArgs(ProgramArgs& args);
+    virtual void initialize(PointTableRef table);
+    virtual void ready(PointTableRef table);
+    virtual void write(const PointViewPtr view);
+    virtual void done(PointTableRef table);
+
+    bool flushCache(size_t size);
+    void addAttribute(draco::GeometryAttribute::Type t, int n);
+    void addGeneric(Dimension::Id pt, int n);
+    void initPointCloud(point_count_t size);
+    void addPoint(int attId, Dimension::IdList idList, PointRef &point, PointId idx);
+    draco::GeometryAttribute::Type getGeometryAttribute(std::string s);
+    void parseDimensions();
+    void parseQuants();
+
+    template <typename T>
+    void addToPointCloud(int attId, Dimension::IdList idList, PointRef &point, PointId idx)
+    {
+        point.setPointId(idx);
+        const auto pointId = draco::PointIndex(idx);
+        //get point information, N dimensional
+        std::vector<T> pointData;
+        for (size_t i = 0; i < idList.size(); ++i) {
+            T data = point.getFieldAs<T>(idList[i]);
+            pointData.push_back(data);
+        }
+
+        draco::PointAttribute *const att = m_pc->attribute(attId);
+        att->SetAttributeValue(att->mapped_index(pointId), pointData.data());
+    }
+
+    struct Args;
+    std::unique_ptr<DracoWriter::Args> m_args;
+
+    std::string m_filename;
+    NL::json m_userDimJson;
+    std::map<Dimension::Id, std::string> m_userDimMap;
+    //these are the default quanitization levels. They will be overridden by any
+    //quantization levels specified in the json argument "quantization"
+    NL::json m_userQuant;
+    std::map<std::string, int> m_quant =
+    {
+        { "POSITION",  11 },
+        { "NORMAL",     7 },
+        { "TEX_COORD", 10 },
+        { "COLOR",      8 },
+        { "GENERIC",    8 }
+    };
+
+    FileStreamPtr m_stream;
+
+    std::map<draco::GeometryAttribute::Type, int> m_dims;
```

Comment:
The purpose of all of these maps is unclear. At a minimum they should be documented here. I think a bunch of this can go away. I think all you're trying to do is the same thing as in the writer - build a single mapping:

```
struct
{
  Dimension::Id;
  Draco Dimension;
  Draco Position; // Red 0, Green 1, Blue 2, for example
  Draco Type;
  Quanitzation;
};
```
Once you've built this you can loop through it once to create the Draco dimensions and then use it when looping through the points to write them to Draco.